components/measure-upload/MeasureRepositoryUpload.tsx

```diff
@@ -0,0 +1,235 @@
+    } else if (measure.name) {
+      label = measure.version ? `${measure.name}|${measure.version}` : measure.name;
+    } else {
+      label = measure.id as string;
+    }
+    displayMap[measure.id as string] = label;
+    return displayMap;
+  }
+
+  /*
+    Sends a `GET` request to the `/Measure` endpoint of the specified Measure Repository Service.
+    Parses and stores the measureIds if successful
+  */
+  async function retrieveMeasures() {
+    setIsLoadingIds(true);
+    try {
+      const response = await fetch(`${measureRepositoryUrl}/Measure`);
+      const responseBody = await response.json();
+      if (response.status === 200) {
+        const displayMap = responseBody.entry.reduce(createMeasureDisplayMap, {});
+        setMeasureBundle(mb => ({ ...mb, displayMap }));
+        if (!measureSelectIsOpen) {
+          setMeasureSelectIsOpen(true);
+        }
+      } else {
+        showNotification({
+          icon: <IconAlertCircle />,
+          title: 'Failed reaching out to server',
+          message: `Server responded with code: ${response.status}:${response.statusText}`,
+          color: 'red'
+        });
+      }
+    } catch (e) {
+      showNotification({
+        icon: <IconAlertCircle />,
+        title: 'Failed reaching out to server',
+        message: `Could not connect to server with url: ${measureRepositoryUrl}`,
+        color: 'red'
+      });
+    }
+    setIsLoadingIds(false);
+  }
+
+  /*
+    Sends a `POST` request to the `Measure/:id/$package` endpoint of the specified Measure
+    Repository Service. Checks the resulting bundle is valid and saves it if so.
+  */
+  async function retrieveMeasurePackage(id: string) {
+    setMeasureBundle(mb => ({ ...mb, selectedMeasureId: id }));
+    if (id) {
+      const display = displayMap[id];
+      setIsLoadingPackage(PACKAGE_STATES.LOADING);
+      const response = await fetch(`${measureRepositoryUrl}/Measure/${id}/$package`, MEASURE_RETRIEVAL_OPTIONS);
+      const responseBody = await response.json();
+      if (responseBody.resourceType !== 'Bundle') {
+        if (responseBody.resourceType === 'OperationOutcome') {
+          rejectUpload(
+            `Packaging of Measure: ${display} failed with message: "${responseBody.issue[0].details.text}"`,
+            display
+          );
+        } else {
+          rejectUpload("Uploaded file must be a JSON FHIR Resource of type 'Bundle'", display);
+        }
+        return;
+      }
+
+      const measures = responseBody.entry.filter((e: fhir4.BundleEntry) => e.resource?.resourceType === 'Measure');
+      if (measures.length !== 1) {
+        rejectUpload(`Expected exactly 1 Measure in uploaded Bundle, received: ${measures.length}`, display);
+        return;
+      }
+
+      const missingValueSets = await identifyMissingValueSets(responseBody);
+      if (missingValueSets.length > 0) {
+        rejectUpload(missingValueSets, display, true);
+        return;
+      }
```

Comment:
If `rejectUpload` and `rejectFile` can be generalized to be the same function, then this could also be I think.